sklearn/preprocessing/tests/test_encoders.py

```diff
@@ -690,9 +660,229 @@ def test_encoders_has_categorical_tags(Encoder):
+
+
+@pytest.mark.parametrize("missing_value", [np.nan, None])
+def test_ohe_missing_value_object_drop(missing_value):
+    # encoder with missing values with object dtypes
+    X = np.array([['a', 'b', missing_value, 'a', missing_value]],
+                 dtype=object).T
+    ohe = OneHotEncoder(sparse=False, drop=[missing_value]).fit(X)
+    assert_array_equal(ohe.drop_idx_, [2])
+
+    X_test = np.array([['a', missing_value, 'b']], dtype=object).T
+    expected_X_trans = np.array([
+        [1, 0],
+        [0, 0],
+        [0, 1]
+    ])
+    X_trans = ohe.transform(X_test)
+    assert_allclose(expected_X_trans, X_trans)
+
+
+@pytest.mark.parametrize("missing_value", [np.nan, None])
+def test_ohe_missing_value_object_auto(missing_value):
+    # encoder with missing values with object dtypes
+
+    X = np.array([['a', 'b', missing_value, 'a', missing_value]],
+                 dtype=object).T
+    ohe = OneHotEncoder(sparse=False, handle_unknown='ignore').fit(X)
+
+    assert len(ohe.categories_) == 1
+
+    if missing_value is None:
+        assert_array_equal(ohe.categories_[0], ['a', 'b', None])
+    else:  # nan
+        assert_array_equal(ohe.categories_[0][:2], ['a', 'b'])
+        assert np.isnan(ohe.categories_[0][-1])
+
+    X_test = np.array([['c', 'a', missing_value, 'b']], dtype=object).T
+    expected_X_trans = np.array([
+        [0, 0, 0],
+        [1, 0, 0],
+        [0, 0, 1],
+        [0, 1, 0]
+    ])
+    X_trans = ohe.transform(X_test)
+    assert_allclose(expected_X_trans, X_trans)
+
+    X_inverse = ohe.inverse_transform(expected_X_trans)
+    expected_X_inverse = np.array([[None, 'a', missing_value, 'b']],
+                                  dtype=object).T
+
+    if missing_value is None:
+        assert_array_equal(X_inverse, expected_X_inverse)
+    else:
+        assert_array_equal(expected_X_inverse[[0, 1, 3], 0],
+                           X_inverse[[0, 1, 3], 0])
+        assert np.isnan(X_inverse[2, 0])
+
+    names = ohe.get_feature_names()
+    assert_array_equal(names, ['x0_a', 'x0_b', f'x0_{missing_value}'])
+
+
+def test_ohe_missing_value_object_different_order():
+    # categories are not in lexical ordering
+    categories = [['b', None, 'a']]
+    X = np.array([['a', 'b', 'b', None, 'c']], dtype=object).T
+    ohe = OneHotEncoder(sparse=False, handle_unknown='ignore',
+                        categories=categories).fit(X)
+
+    X_test = np.array([['a', 'b', None, 'c']], dtype=object).T
+    X_trans = ohe.fit_transform(X_test)
+    expected_X_trans = np.array([
+        [0, 0, 1],
+        [1, 0, 0],
+        [0, 1, 0],
+        [0, 0, 0]
+    ])
+    assert_allclose(expected_X_trans, X_trans)
+
+
+@pytest.mark.parametrize("categories", [
+    'auto', [['a', 'b', None], [0, 2, np.nan]]
+])
+def test_ohe_missing_value_mixed_sanity(categories):
+    # sanity check for mixed data with missing values
+    X = np.array([['a', 'b', None], [0, np.nan, 2]], dtype=object).T
```

Comment:
I find the use of `.T` with multiple samples and features easy to miss... maybe use .transpose() instead, or avoid it.